Add channel shuffle op fp32 + quantized.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36815 Add channel shuffle op fp32 + quantized.**
* #36814 Replace empty_affine_quantizer with direct dispatch to at::native::empty_affine.. .
* #36813 Add quantized adaptive avgpool.
* #36812 Move to using MemoryFormat::ChannelsLast for quantized avgpool2d.

Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

kimishpatel committed:
Add channel shuffle op fp32 + quantized.

Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

dreiss commented on aten/src/ATen/native/ChanelShuffle.cpp:
```diff
@@ -0,0 +1,44 @@
+      "Number of groups to divide channels in must be positive.",
+      " Value of groups:", groups);
+  AT_ASSERTM((c % groups) == 0,
+             "Number of channels must be divisible gy groups. Got ",
```
"gy" -> "by"

dreiss commented on aten/src/ATen/native/ChanelShuffle.cpp:
```diff
@@ -0,0 +1,44 @@
+  // For ChannelsFirst, a.k.a Contiguous, memory format we will also need
+  // a fast custom implementation perhaps.
+  return input_reshaped.permute({0 /* b */, 2 /* oc */, 1 /* groups */, 3})
+                       .contiguous()
```
Why do we need to call contiguous at all here?

kimishpatel replied:
So this will actually rearrange the channels, as we need to reshape it back to the original shape. Was that your question?

dreiss replied:
What happens if you remove this contiguous call entirely?  What happens if you move it after the reshape and give it the `input.suggest_input_format()` argument?

kimishpatel replied:
For removing `contiguous` entirely: This depends how will the strides be computed since reshape's input shape will be post permute and how might actually try to map part of permuted channels to h and w dims. I think. It is not clear without looking at the details of restriding logic.
For the second question my response is the same. 
In the current form it says. If you have 16 channels that you divide in 2 groups each of 8 channels, rearrange as follows: All groups of a particular channel are together. Thus 2 groups of channel 0, 2 groups of channel 1.... 2 groups of channel 7. Having asked this kind of permute of channels, it seems cleaner to call contiguous on that so that all groups are actually contiguous in memory and then you reshape it back to original shape. Now you get the same number of original channels but arranged differently. Doing it other ways your question asked, requires to understand how restriding is done.

dreiss commented on aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:
```diff
@@ -0,0 +1,115 @@
+      createStatus == pytorch_qnnp_status_success,
+      "failed to create QNNPACK ChannelShuffle operator");
+
+  std::unique_ptr<pytorch_qnnp_operator, QnnpackOperatorDeleter>
+      qnnpack_uniq_ptr(qnnpack_operator);
```
I commented on the old diff as well, but this style ensure that the operator is freed even if we throw an exception.

dreiss commented on aten/src/ATen/native/ChanelShuffle.cpp:
```diff
@@ -0,0 +1,44 @@
+
+  auto input_reshaped = self.view({b, groups, oc, -1});
+  // TODO: contiguous can be made to preserve the memory format
+  // of the input. However since the above reshape clobbers h and w
+  // it may not be safe to do that, since channels_last contiguous
+  // may think oc and and the last dim correspond to h,w?
```
Yeah, I think this is right.

dreiss commented on aten/src/ATen/native/ChanelShuffle.cpp:
```diff
@@ -0,0 +1,44 @@
+  // a fast custom implementation perhaps.
+  return input_reshaped.permute({0 /* b */, 2 /* oc */, 1 /* groups */, 3})
+                       .contiguous()
+                       .reshape(self.sizes());
```
After calling reshape, can we also try to preserve the dimension names from the input?

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]

kimishpatel committed:
Update on "Add channel shuffle op fp32 + quantized."


Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Differential Revision: [D21093841](https://our.internmc.facebook.com/intern/diff/D21093841/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21093841/)!

[ghstack-poisoned]
